conda_libmamba_solver/solver.py

```diff
@@ -256,18 +258,44 @@ def _spinner_msg_solving(self):
                 return "Getting pinned dependencies"
         return "Solving environment"
 
+    def _max_attempts(self, in_state: SolverInputState):
+        from_env_var = os.environ.get("CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS")
+        n_installed = len(in_state.installed)
+        if from_env_var:
+            try:
+                max_attempts_from_env = int(from_env_var)
+            except ValueError:
+                raise CondaValueError(
+                    f"CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS='{from_env_var}'. Must be int."
+                )
+            if max_attempts_from_env < 1:
+                raise CondaValueError(
+                    f"CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS='{max_attempts_from_env}'. Must be >=1."
+                )
+            if max_attempts_from_env > n_installed:
+                log.warning(
+                    "CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS='%s' is higher than the number of "
+                    "installed packages (%s). Using that one instead.",
+                    max_attempts_from_env,
+                    n_installed,
+                )
+                return n_installed
+            return max_attempts_from_env
+        if in_state.update_modifier.FREEZE_INSTALLED:
+            # this the default, but can be overriden with --update-specs
+            # we cap at MAX_SOLVER_ATTEMPTS_CAP attempts to avoid things
+            # getting too slow in large environments
+            return min(self.MAX_SOLVER_ATTEMPTS_CAP, n_installed)
+        return 1
+
     def _solving_loop(
         self,
         in_state: SolverInputState,
         out_state: SolverOutputState,
         index: LibMambaIndexHelper,
     ):
         solved = False
-        max_attempts = max(
-            2,
-            int(os.environ.get("CONDA_LIBMAMBA_SOLVER_MAX_ATTEMPTS", len(in_state.installed))) + 1,
-        )
-        for attempt in range(1, max_attempts):
+        for attempt in range(1, self._max_attempts(in_state) + 1):
```

Comment:
Would it be useful to prepopulate the max attempts at the start of the solve as a instance attribute, instead of inline, so it's easier to debug?

Reply:
I'm not sure if this might break some tests. It shouldn't, but by saving it as instance attribute we are assuming we are going to have the same number of attempts for every call to `solve_for_*()` (for the lifetime of the same instance). In some instances, that might be tied to `len(in_state.installed)`.

So I'd rather keep it like it is because it's technically more correct and does not assume that "one instantiation, one solve".

Reply:
Sounds good, thank you!